src/runtime/dynlink/src/engines/twizzler.rs

```diff
@@ -0,0 +1,168 @@
+
+use crate::{
+    compartment::CompartmentId,
+    context::engine::{ContextEngine, LoadDirective, LoadFlags},
+    library::BackingData,
+    DynlinkError, DynlinkErrorKind,
+};
+
+#[derive(Clone)]
+pub struct Backing {
+    obj: twizzler_runtime_api::ObjectHandle,
+}
+
+impl Backing {
+    pub fn new(inner: twizzler_runtime_api::ObjectHandle) -> Self {
+        Self { obj: inner }
+    }
+}
+
+impl BackingData for Backing {
+    fn data(&self) -> (*mut u8, usize) {
+        (
+            unsafe { self.obj.start.add(NULLPAGE_SIZE) },
+            // one null-page, one meta-page
+            MAX_SIZE - NULLPAGE_SIZE * 2,
+        )
+    }
+
+    fn new_data() -> Result<Self, DynlinkError> {
+        let runtime = twizzler_runtime_api::get_runtime();
+        let id = sys_object_create(
+            ObjectCreate::new(
+                BackingType::Normal,
+                LifetimeType::Volatile,
+                None,
+                ObjectCreateFlags::empty(),
+            ),
+            &[],
+            &[],
+        )
+        .map_err(|_| DynlinkErrorKind::NewBackingFail)?;
+        let handle = runtime
+            .map_object(id.as_u128(), MapFlags::READ | MapFlags::WRITE)
+            .map_err(|_| DynlinkErrorKind::NewBackingFail)?;
+        Ok(Self { obj: handle })
+    }
+
+    fn load_addr(&self) -> usize {
+        self.obj.start as usize
+    }
+
+    type InnerType = twizzler_runtime_api::ObjectHandle;
+
+    fn to_inner(self) -> Self::InnerType {
+        self.obj
+    }
+}
+
+pub struct Engine;
+
+fn within_object(slot: usize, addr: usize) -> bool {
+    addr >= slot * MAX_SIZE + NULLPAGE_SIZE && addr < (slot + 1) * MAX_SIZE - NULLPAGE_SIZE * 2
+}
+
+impl ContextEngine for Engine {
+    type Backing = Backing;
+
+    fn load_segments(
+        &mut self,
+        src: &Self::Backing,
+        ld: &[crate::context::engine::LoadDirective],
+    ) -> Result<Vec<Self::Backing>, DynlinkError> {
+        let create_spec = ObjectCreate::new(
+            BackingType::Normal,
+            LifetimeType::Volatile,
+            None,
+            ObjectCreateFlags::empty(),
+        );
+
+        let build_copy_cmd = |directive: &LoadDirective| {
+            if !within_object(
+                if directive.load_flags.contains(LoadFlags::TARGETS_DATA) {
+                    1
+                } else {
+                    0
+                },
+                directive.vaddr,
+            ) || directive.memsz > MAX_SIZE - NULLPAGE_SIZE * 2
+                || directive.offset > MAX_SIZE - NULLPAGE_SIZE * 2
+                || directive.filesz > directive.memsz
+            {
+                return Err(DynlinkError::new(DynlinkErrorKind::LoadDirectiveFail {
+                    dir: *directive,
+                }));
+            }
+
+            if directive.filesz != directive.memsz {
+                todo!()
+            }
+
+            let src_start = (NULLPAGE_SIZE + directive.offset) & !(directive.align - 1);
```

Comment:
do we need this alignment for `src_start`?

Reply:
I think we need it for either both, or neither. I think that now the way copy-from works on unaligned addresses, we probably could remove the alignment stuff, but this will probably make directives be page-aligned, which will reduce copying in the kernel.

Unless I've missed something. Most ELF loaders I've looked at use the alignment in roughly this way.